main/modes/games/ultimateTTT/ultimateTTTcpuPlayer.c

```diff
@@ -0,0 +1,1192 @@
+    int playerCount         = 0;
+    int opponentCount       = 0;
+    tttPlayer_t otherPlayer = (player == TTT_P1) ? TTT_P2 : TTT_P1;
+
+    for (int y = 0; y < 3; y++)
+    {
+        if (game[x][y] == TTT_NONE)
+        {
+            noneCount++;
+        }
+        else if (game[x][y] == player)
+        {
+            playerCount++;
+        }
+        else if (game[x][y] == otherPlayer)
+        {
+            opponentCount++;
+        }
+    }
+
+    return (rowCount_t)(playerCount | opponentCount << 2 | noneCount << 4);
+}
+
+static rowCount_t checkDiag(const tttPlayer_t game[3][3], int n, tttPlayer_t player)
+{
+    int noneCount           = 0;
+    int playerCount         = 0;
+    int opponentCount       = 0;
+    tttPlayer_t otherPlayer = (player == TTT_P1) ? TTT_P2 : TTT_P1;
+
+    if (n == 1)
+    {
+        for (int i = 0; i < 3; i++)
+        {
+            int x = i;
+            int y = (n == 1) ? i : (2 - i);
+
+            if (game[x][y] == TTT_NONE)
+            {
+                noneCount++;
+            }
+            else if (game[x][y] == player)
+            {
+                playerCount++;
+            }
+            else if (game[x][y] == otherPlayer)
+            {
+                opponentCount++;
+            }
+        }
+    }
+
+    return (rowCount_t)(playerCount | opponentCount << 2 | noneCount << 4);
+}
+
+// So naively scoring based on just the moves to make isn't enough
+// - Move hierarchy doesn't sufficiently capture the value of a move
+// - You have to think about not just how that move advances, but also how many moves away from winning we are
+static uint16_t movesToWin(const tttPlayer_t subgame[3][3], tttPlayer_t player)
+{
+    uint16_t movesToWin = 0;
+    tttPlayer_t result  = tttCheckWinner(subgame);
+    tttPlayer_t simulation[3][3];
+
+    memcpy(simulation, subgame, sizeof(simulation));
+
+#ifdef DEBUG_TTT_CPU
+    TCPU_LOG(">>> Moves to win game:\n");
+    printGame(subgame);
+    TCPU_LOG("---\n");
+#endif
+
+    uint16_t minMovesToWin = 5;
+
+    for (int y = 0; y < 3; y++)
+    {
+        rowCount_t rowCount  = checkRow(subgame, y, player);
+        uint16_t movesNeeded = 3 - TO_PLAYERS(rowCount);
+        if (0 == TO_OPPONENTS(rowCount) && movesNeeded < minMovesToWin)
+        {
+            minMovesToWin = movesNeeded;
+        }
+    }
+
+    for (int x = 0; x < 3; x++)
+    {
+        rowCount_t rowCount  = checkCol(subgame, x, player);
+        uint16_t movesNeeded = 3 - TO_PLAYERS(rowCount);
+        if (0 == TO_OPPONENTS(rowCount) && movesNeeded < minMovesToWin)
+        {
+            minMovesToWin = movesNeeded;
+        }
+    }
+
+    for (int n = 0; n < 2; n++)
+    {
+        rowCount_t rowCount  = checkDiag(subgame, n, player);
+        uint16_t movesNeeded = 3 - TO_PLAYERS(rowCount);
+        if (0 == TO_OPPONENTS(rowCount) && movesNeeded < minMovesToWin)
+        {
+            minMovesToWin = movesNeeded;
+        }
+    }
+
+    return minMovesToWin;
```

Comment:
Code after this is dead. Intentional?

Reply:
Ah, no I should delete that. Surprised it doesn't generate a warning

Reply:
Yeah, not even `cppcheck` caught it. Odd.